sw/device/lib/base/print.c

```diff
@@ -0,0 +1,339 @@
+  width = width > kWordBits ? kWordBits : width;
+  while (len < width) {
+    buffer[kWordBits - len - 1] = '0';
+    ++len;
+  }
+  return out.sink(out.data, buffer + (kWordBits - len), len);
+}
+
+/**
+ * Prints out the next entry in |args| according to |spec|.
+ *
+ * This function assumes that |spec| accurately describes the next entry in
+ * |args|.
+ *
+ * @param out the sink to write bytes to.
+ * @param spec the specifier to use for stringifying.
+ * @param bytes_written out param for the number of bytes writen to |out|.
+ * @param va_list the list to pull an entry from.
+ */
+static void process_specifier(buffer_sink_t out, format_specifier_t spec,
+                              size_t *bytes_written, va_list args) {
+  // Switch on the specifier. At this point, we assert that there is
+  // an initialized value of correct type in the VA list; if it is
+  // missing, the caller has caused UB.
+  switch (spec.type) {
+    case kPercent: {
+      *bytes_written += out.sink(out.data, "%", 1);
+      break;
+    }
+    case kCharacter: {
+      char value = (char)va_arg(args, uint32_t);
+      *bytes_written += out.sink(out.data, &value, 1);
+      break;
+    }
+    case kString: {
+      char *value = va_arg(args, char *);
+      size_t len = 0;
+      while (value[len] != '\0') {
+        ++len;
+      }
+      *bytes_written += out.sink(out.data, value, len);
+      break;
+    }
+    case kSizedStr: {
+      size_t len = va_arg(args, uint32_t);
+      char *value = va_arg(args, char *);
+      *bytes_written += out.sink(out.data, value, len);
+      break;
+    }
+    case kSignedDec1:
+    case kSignedDec2: {
+      uint32_t value = va_arg(args, uint32_t);
+      if (((int32_t)value) < 0) {
+        *bytes_written += out.sink(out.data, "-", 1);
+        value = -value;
+      }
+      *bytes_written += write_digits(out, value, spec.width, 10, kDigitsLow);
+      break;
+    }
+    case kUnsignedOct: {
+      uint32_t value = va_arg(args, uint32_t);
+      *bytes_written += write_digits(out, value, spec.width, 8, kDigitsLow);
+      break;
+    }
+    case kPointer: {
+      *bytes_written += out.sink(out.data, "0x", 2);
+      spec.width = 8;
+      // Explicit fallthrough.
+    }
+    case kSvHexLow:
+    case kUnsignedHexLow: {
+      uint32_t value = va_arg(args, uint32_t);
+      *bytes_written += write_digits(out, value, spec.width, 16, kDigitsLow);
+      break;
+    }
+    case kSvHexHigh:
+    case kUnsignedHexHigh: {
+      uint32_t value = va_arg(args, uint32_t);
+      *bytes_written += write_digits(out, value, spec.width, 16, kDigitsHigh);
+      break;
+    }
+    case kUnsignedDec: {
+      uint32_t value = va_arg(args, uint32_t);
+      *bytes_written += write_digits(out, value, spec.width, 10, kDigitsLow);
+      break;
+    }
+    case kSvBinary: {
+      uint32_t value = va_arg(args, uint32_t);
+      *bytes_written += write_digits(out, value, spec.width, 2, kDigitsLow);
+      break;
+    }
+    case kCharsSoFar: {
```

Comment:
This isn't the libc behavior of `%n`.  In libc it causes the _argument_ to get written with the number of characters written so far.  It's been a traditional giant source of security nightmares.  I would completely drop `%n`.

Reply:
Wholeheartedly agree with removing `%n` if this is an option for us.

Reply:
Oh, that's horrific. Thanks for pointing that out.